### PR TITLE
Fix drag annotation blink issue

### PR DESCRIPTION
**Pull request by Chaoba**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Fix drag annotation blink issue</changelog>`.

### Summary of changes
In this pr, we will remove the drag annotation from the drag source only when the map is idle and 1 sec later, so there will be no blink issue after drag.

Before:

https://user-images.githubusercontent.com/8577318/130190854-a89298ce-b63d-4b77-a6d6-29d754885c5e.mp4

After:

https://user-images.githubusercontent.com/8577318/133201118-aa534c68-1026-412a-be75-d981fb66223b.mp4

### CODE REVIEW

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 1]
```diff
@@ -663,16 +664,29 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
...
+        mapListenerDelegate.removeOnMapIdleListener(this)
+        // Remove dragging annotation from drag layer
+        Handler(Looper.getMainLooper()).postDelayed(
+          {
+            // Delay 1 second to avoid blink issue.
+            delegateProvider.getStyle { style ->
+              updateDragSource(style)
+            }
+          },
+          1000
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:678]: move to const

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 2]
```diff
@@ -663,16 +664,29 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
       return false
     }
 
+    private val onMapIdleListener: OnMapIdleListener = object : OnMapIdleListener {
+      override fun onMapIdle() {
+        mapListenerDelegate.removeOnMapIdleListener(this)
+        // Remove dragging annotation from drag layer
+        Handler(Looper.getMainLooper()).postDelayed(
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:671]: could create handler only once and store it as a val

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 1 above)
Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:678]: and why 1000? isn't it too much?

**Comment by kiryldz** [comment]
@Chaoba `Fix drag annotation blink; Apply properties to drag layer` is pretty confusing imho.
Please better describe how it was fixed stating that we introduce delay when drag is finished before removing drag source.
Also could you please clarify if all those tests updates are related to drag blink issue? if not - can we move it to a separate PR?

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 1 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:678]: Users can't notice this delay, we need this buffer to cover the update delay, if too short, the blink may come back when there are too many annotations.

**Comment by kiryldz** [comment]
Also when I look at the second video when you start dragging annotation again (sec 4-5) it seems that drag start is not behaving as expected, annotation jumps to some position instead of smooth dragging.

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 2 above)
Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:671]: also what if user starts dragging again while we're waiting for one second here?
shouldn't we do `handler.removeCallbackAndMessages(null)` when drag starts?

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 1 above)
Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:678]: Well, short is a pretty common word 😄 We don't know what device / how many annotations user is adding. If it's the only way to go do we need to make that delay configurable by adding getter / setter from the user side?

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 2 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt]: If users drag again, `draggingAnnotation` will not be null, that's simply what we do while dragging. You can see from the attached video, the time between two drags is obviously less than 1 second and it works well.

Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt]: The problem that it does not 😄 
https://github.com/mapbox/mapbox-maps-android/pull/639#issuecomment-918855680

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 1 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt]: IMO, we should expose this config to users unless they have a strong request on it. Would love to get enough feedback before that step.

**Comment by Chaoba** [comment]
> Also when I look at the second video when you start dragging annotation again (sec 4-5) it seems that drag start is not behaving as expected, annotation jumps to some position instead of smooth dragging.

We still need to remove the dragging annotation from original source, which will trigger the update for all annotations. We can't skip this step in the current architecture. What we currently do is making it better but not the best.

**Comment by kiryldz** [comment]
> > Also when I look at the second video when you start dragging annotation again (sec 4-5) it seems that drag start is not behaving as expected, annotation jumps to some position instead of smooth dragging.
> 
> We still need to remove the dragging annotation from original source, which will trigger the update for all annotations. We can't skip this step in the current architecture. What we currently do is making it better but not the best.

Why is there no such problem when the very first dragging happens?

**Comment by pengdev** [comment]
> > > Also when I look at the second video when you start dragging annotation again (sec 4-5) it seems that drag start is not behaving as expected, annotation jumps to some position instead of smooth dragging.
> > 
> > 
> > We still need to remove the dragging annotation from original source, which will trigger the update for all annotations. We can't skip this step in the current architecture. What we currently do is making it better but not the best.
> 
> Why is there no such problem when the very first dragging happens?

I took a closer look at the first video, I see the annotation being dragged also jumped at the beginning, instead of smooth transition, so in this regard I don't think the behaviour is a regression compared to main branch.

**Comment by Chaoba** [comment]
> I took a closer look at the first video, I see the annotation being dragged also jumped at the beginning, instead of smooth transition, so in this regard I don't think the behaviour is a regression compared to main branch.

Exactly. We have to update all annotations in order to remove the dragging annotation from the original source at the beginning of the drag gesture. Currently we don't have a solution for it.

**Comment by pengdev** [comment]
@Chaoba one thought, can we keep both layers during the transition? for example, start merging the dragged annotation to the annotation source, and listen to the MapIdle event, when the MapIdle event is triggered, then remove the dragged layer, wdyt?

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 3]
```diff
@@ -663,16 +665,29 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
...
     }
 
+    private val onMapIdleListener: OnMapIdleListener = object : OnMapIdleListener {
+      override fun onMapIdle() {
+        mapListenerDelegate.removeOnMapIdleListener(this)
+        // Remove dragging annotation from drag layer
+        handler.postDelayed(
+          {
+            // Delay 1 second to avoid blink issue.
+            delegateProvider.getStyle { style ->
```
Comment (alexander-kulikovskii) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:675]: can we get crash when remove (set to null) annotation manager before this postDelayed happen?
should we remove it from handler?

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:675]: Not see a crash when test this scenario locally. But need remove it in `onDestroy`.

**Comment by Chaoba** [comment]
> @Chaoba one thought, can we keep both layers during the transition? for example, start merging the dragged annotation to the annotation source, and listen to the MapIdle event, when the MapIdle event is triggered, then remove the dragged layer, wdyt?

That's what we are doing now. But we always keep two layers.

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 4]
```diff
@@ -663,16 +666,29 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
       return false
     }
 
+    private val onMapIdleListener: OnMapIdleListener = object : OnMapIdleListener {
+      override fun onMapIdle() {
+        mapListenerDelegate.removeOnMapIdleListener(this)
+        // Remove dragging annotation from drag layer
+        handler.postDelayed(
+          {
+            // Delay 1 second to avoid blink issue.
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:675]: Delay is now 3 seconds isn't it?

**Review by kiryldz — APPROVED** [review]

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 4 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:675]: This is used to test the handler after onDestroy. Still need to keep it 1 second.

**Review by alexander-kulikovskii — APPROVED** [review]

**Pull request merged**